src/terms/local.jl

```diff
@@ -136,28 +153,33 @@ function compute_forces(::TermAtomicLocal, basis::PlaneWaveBasis{T}, ψ, occupat
 end
 @timing "forces: local" function forces_local(S, basis::PlaneWaveBasis{T}, ρ, q) where {T}
     model = basis.model
-    recip_lattice = model.recip_lattice
-    ρ_fourier = to_cpu(fft(basis, total_density(ρ)))
     real_ifSreal = S <: Real ? real : identity
 
-    # TODO: currently, local forces entierly computed on the CPU.
-    #       This might not be optimal. See compute_local_potential() too.
     Gqs_cart = [model.recip_lattice * (G + q) for G in to_cpu(G_vectors(basis))]
-    form_factors = atomic_local_form_factors(basis, Gqs_cart)
+    form_factors, iG2norm = atomic_local_form_factors(basis, Gqs_cart)
+
+    Gqs = map(G -> G+q, G_vectors(basis))
+    ρ_fourier = reshape(fft(basis, total_density(ρ)), length(Gqs))
+
+    indices = to_device(basis.architecture, collect(1:length(Gqs)))
+    ρpot = similar(ρ_fourier)
+    twoπp = similar(ρ_fourier)
 
     # energy = sum of form_factor(G) * struct_factor(G) * rho(G)
     # where struct_factor(G) = e^{-i G·r}
     forces = [zero(Vec3{S}) for _ = 1:length(model.positions)]
     for (igroup, group) in enumerate(model.atom_groups)
-        element = model.atoms[first(group)]
         for idx in group
             r = model.positions[idx]
-            forces[idx] = -real_ifSreal(sum(conj(ρ_fourier[iG])
-                                              * form_factors[(igroup, norm(Gqs_cart[iG]))]
-                                              * cis2pi(-dot(G + q, r))
-                                              * (-2T(π)) * (G + q) * im
-                                              / sqrt(model.unit_cell_volume)
-                                        for (iG, G) in enumerate(to_cpu(G_vectors(basis)))))
+
```

Comment:
This new code is generally harder to read than what we had. To what extend is this needed to get it to work / performance ?

Reply:
I agree the code loses in clarity. The main challenge for the GPU is to use compatible constructs, and an explicit for loop is not one of them. My hope is that the comment describing the formula before the loop is enough to guide the reader. If not, an additional comment could be added before each `map!` describing what it does

Reply:
Makes sense. Yes, I think the comment helps.